ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,249 @@ func (ps *prioritySampler) apply(spn *span) {
 	}
 	spn.SetTag(keySamplingPriorityRate, rate)
 }
+
+// rulesSampler allows a user-defined list of rules to apply to spans.
+// These rules can match based on the span's Service, Name or both.
+// When making a sampling decision, the rules are checked in order until
+// a match is found.
+// If a match is found, the rate from that rule is used.
+// If no match is found, and the DD_TRACE_SAMPLE_RATE environment variable
+// was set to a valid rate, that value is used.
+// Otherwise, the rules sampler didn't apply to the span, and the decision
+// is passed to the priority sampler.
+//
+// The rate is used to determine if the span should be sampled, but an upper
+// limit can be defined using the DD_TRACE_RATE_LIMIT environment variable.
+// Its value is the number of spans to sample per second.
+// Spans that matched the rules but exceeded the rate limit are not sampled.
+type rulesSampler struct {
+	rules   []SamplingRule
+	rate    float64
+	limiter *rate.Limiter
+
+	// "effective rate" calculations
+	mu           sync.Mutex // guards below fields
+	ts           time.Time  // timestamp, to detect when counters need resetting
+	allowed      int        // number of spans allowed by rate limiter since ts
+	seen         int        // number of spans checked by rate limiter since ts
+	previousRate float64    // previous second's rate, averaged with current rate for smoothing
+}
+
+// newRulesSampler configures a *rulesSampler instance using the given set of rules.
+// Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
+func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	rate := sampleRate()
+	return &rulesSampler{
+		rules:   appliedSamplingRules(rules),
+		rate:    rate,
+		limiter: newRateLimiter(rate),
+		ts:      time.Now().Truncate(time.Second),
+	}
+}
+
+// appliedSamplingRules validates the user-provided rules and returns an internal representation.
+// If the DD_TRACE_SAMPLING_RULES environment variable is set, it will replace the given rules.
+func appliedSamplingRules(rules []SamplingRule) []SamplingRule {
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		rules = rules[:0]
+		jsonRules := []struct {
+			Service string      `json:"service"`
+			Name    string      `json:"name"`
+			Rate    json.Number `json:"rate"`
```

Comment:
After discussion with other languages and in order to provide consistency, please change this key from `rate` to `sample_rate`.